This patch will no longer be necessary in 3.2
See GNOME #658176

Index: gnome-shell-3.0.2/src/shell-app-system.c
===================================================================
--- gnome-shell-3.0.2.orig/src/shell-app-system.c	2011-09-04 14:57:31.574689356 +0200
+++ gnome-shell-3.0.2/src/shell-app-system.c	2011-09-04 14:59:17.555207468 +0200
@@ -195,12 +195,9 @@ shell_app_system_init (ShellAppSystem *s
   /* Key is owned by info */
   priv->app_id_to_app = g_hash_table_new (g_str_hash, g_str_equal);
 
-  /* For now, we want to pick up Evince, Nautilus, etc.  We'll
-   * handle NODISPLAY semantics at a higher level or investigate them
-   * case by case.
-   */
-  priv->apps_tree = gmenu_tree_lookup ("applications.menu", GMENU_TREE_FLAGS_INCLUDE_NODISPLAY);
-  priv->settings_tree = gmenu_tree_lookup ("gnomecc.menu", GMENU_TREE_FLAGS_INCLUDE_NODISPLAY);
+  /* Do not show NODISPLAY menus and applications */
+  priv->apps_tree = gmenu_tree_lookup ("applications.menu", GMENU_TREE_FLAGS_NONE);
+  priv->settings_tree = gmenu_tree_lookup ("gnomecc.menu", GMENU_TREE_FLAGS_NONE);
 
   priv->app_change_timeout_id = 0;
 
